jupyterlite/jupyterlite
# Add localforage memory fallback

## References

- fixes #535
- expands on #525

## Code changes

- [x] extracts the localforage singleton to a plugin
- [x] adds a plugin for memory storage
- [x] adds a `enableMemoryStorage` config value to enable memory storage
  - shipping default `disabledPlugins` seems... bad  
  - was enabled by default on #525, but can-lose-data-by-default is not great
- [x] adds `ready` promises to settings and contents
  - this allows enough time for extra drivers to be installed
- [x] add `(settings|contents)StorageDrivers` config value
  - if left as the default (`null`), will fall back to the "best available" (apparently `IndexedDB`)
- [x] disable persistence on `repl` in docs app 
- [x] update configuring docs

## User-facing changes

- iframe-embedded apps, e.g. `repl` can be configured without extra `sandbox` permissions and still work
- while the as-distributed `repl` will have persistent settings...
  - on _our_ docs site, settings will _not_ persist between page loads
    - this appears to impact theme/scrollbars/font size... kind of annoying to lose, but ultimately `repl` is more about _publisher_ control, anyway, and we need to keep this feature under test/demonstration

## Backwards-incompatible changes

- adds ready promise to `IContents` and `ISettings`
- shortened some `(Settings|Contents).Options` members (remove the `settings` and `contents` prefixes)

## Test Procedure

- Go to the RTD preview
- See lab
  - in the _Application_ tab in dev tools, see the IndexedDB
- in the same tab, go to the repl app
  - delete the `JupyterLite` IndexedDB
- choose a different theme through `Ctrl+Shift+C`modal
- reload the repl, see that the theme reverts to the default
- see that no new IndexedDB is created

## Future Work

- there are a number of other localforage drivers
  - some of these might be interesting to explore as extensions
    - these could add the drivers and subclass the existing settings and contents, and overload the `createDefault*Storage` methods to create interesting effects
  - ...but probably don't need to go in core
